solutions/gold/kattis-playlist.mdx

```diff
@@ -0,0 +1,165 @@
+$\binom{100}{4} \cdot 100 = 4 \cdot 10^8$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N(\binom{N}{4}+N))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <algorithm>
+#include <array>
+#include <iostream>
+#include <numeric>
+#include <unordered_map>
+#include <vector>
+
+using namespace std;
+
+typedef array<char, 4> State;
+
+unordered_map<string, int> artistmap;
+unordered_map<long long, int> from[2][100];
+vector<State> states[2][100];
+vector<int> g[2][100];
+vector<int> names;
+
+long long getid(const State &state) {
+	long long id = 0;
+	for (int i = 0; i < 4; i++) { id |= ((long long)state[i] << (i * 8)); }
+	return id;
+};
+
+void dfs(State state, int node, int dist, int rev) {
+	if (dist == 4) { return; }
+	state[0] = names[node];
+	sort(state.begin(), state.end());
+	long long id = getid(state);
+	for (int son : g[rev][node]) {
+		if (find(state.begin(), state.end(), names[son]) == state.end() &&
+		    !from[rev][son].count(id)) {
+			if (dist == 3) { states[rev][son].push_back(state); }
+			from[rev][son][id] = node;
+			dfs(state, son, dist + 1, rev);
+		}
+	}
+};
+
+void reconstruct(State state, int node, bool rev) {
+	vector<int> path;
+	for (int i = 0; i < 4; i++) {
+		int parent = from[rev][node][getid(state)];
+		for (int j = 0; j < 4; j++) {
+			if (state[j] == names[parent]) { state[j] = 0; }
+		}
+		sort(state.begin(), state.end());
+		path.push_back(parent + 1);
+		node = parent;
+	}
+	if (!rev) { reverse(path.begin(), path.end()); }
+	for (int node : path) { cout << node << ' '; }
+};
+
+int main() {
+	int n;
+	cin >> n;
+
+	names.resize(n);
+	for (int i = 0; i < n; i++) {
+		string name;
+		cin >> name;
+		// Associate to each artist a number
+		if (!artistmap.count(name)) { artistmap[name] = artistmap.size() + 1; }
+		names[i] = artistmap[name];
+		int k;
+		cin >> k;
+		for (int j = 0; j < k; j++) {
+			int node;
+			cin >> node;
+			node--;
+			// If the next sogn has the same artist, it won't be considered
+			if (names[i] != names[node]) {
+				g[0][i].push_back(node);
+				g[1][node].push_back(i);
+			}
+		}
+	}
+
+	// Compute the 4-length chains starting from node i
+	for (int i = 0; i < n; i++) {
+		dfs({}, i, 0, 0);
+		dfs({}, i, 0, 1);
+	}
+
+	for (int i = 0; i < n; i++) {
+		unordered_map<long long, int> m;
+		for (State state : states[0][i]) {
+			for (int j = 0; j < 16; j++) {
+				auto s = state;
+				for (int k = 0; k < 4; k++) {
+					if (j & (1 << k)) { s[k] = 0; }
+				}
+				sort(s.begin(), s.end());
+				m[getid(s)]++;
+			}
+		}
+		for (State state : states[1][i]) {
+			int cnt = (int)states[0][i].size();
+			for (int j = 0; j < 15; j++) {
+				auto s = state;
+				int odd = 0;
+				for (int k = 0; k < 4; k++) {
+					if (j & (1 << k)) {
+						s[k] = 0;
+						odd ^= 1;
+					}
+				}
+				sort(s.begin(), s.end());
+				// Inclusion-Exclusion on sets
+				cnt += (1 - 2 * odd) * m[getid(s)];
+			}
+			if (cnt) {
+				for (State t : states[0][i]) {
```

Comment:
when you're looping thru the states, do they have to be copies? or can they be const refs?

Reply:
No they don't have.

Reply:
Done, I made all const refs